### PR TITLE
NXDRIVE-1109: Use the pathlib module to handle paths

### PR DESCRIPTION
**Pull request by LKleinNux**

### CODE REVIEW

File: tests/test_watchers.py [hunk 1]
```diff
@@ -352,11 +353,11 @@ def test_watcher_remote_id_setter(self):
         # Create files with Unicode combining accents,
         # Unicode latin characters and no special characters
         file_path = local.abspath("/Test.pdf")
-        copyfile(self.location + "/resources/testFile.pdf", file_path)
+        copyfile(self.location / "resources/testFile.pdf", file_path)
```
Comment (BoboTiG) [inline tests/test_watchers.py]: `copyfile(self.location / "resources" / "testFile.pdf", file_path)`

File: tests/test_watchers.py [hunk 2]
```diff
@@ -352,11 +353,11 @@ def test_watcher_remote_id_setter(self):
...
         # Wait for test workspace synchronization
         self.wait_sync()
         remote_id = local.get_remote_id("/Test.pdf")
-        copyfile(self.location + "/resources/testFile.pdf", file_path)
+        copyfile(self.location / "resources/testFile.pdf", file_path)
```
Comment (BoboTiG) [inline tests/test_watchers.py]: `copyfile(self.location / "resources" / "testFile.pdf", file_path)`

File: tests/test_watchers.py [hunk 3]
```diff
@@ -367,13 +368,13 @@ def test_watcher_remote_id_setter_stopped(self):
         self.wait_sync()
 
         file_path = self.local_1.abspath("/Test.pdf")
-        copyfile(self.location + "/resources/testFile.pdf", file_path)
+        copyfile(self.location / "resources/testFile.pdf", file_path)
```
Comment (BoboTiG) [inline tests/test_watchers.py]: `copyfile(self.location / "resources" / "testFile.pdf", file_path)`

File: tests/test_watchers.py [hunk 4]
```diff
@@ -367,13 +368,13 @@ def test_watcher_remote_id_setter_stopped(self):
...
         # Give some time for the local watcher to handle the copy
         sleep(5)
 
         self.engine_1.stop()
         remote_id = self.local_1.get_remote_id("/Test.pdf")
-        copyfile(self.location + "/resources/testFile.pdf", file_path)
+        copyfile(self.location / "resources/testFile.pdf", file_path)
```
Comment (BoboTiG) [inline tests/test_watchers.py]: `copyfile(self.location / "resources" / "testFile.pdf", file_path)`

File: nxdrive/constants.py [hunk 5]
```diff
@@ -16,6 +17,7 @@
 
 DOWNLOAD_TMP_FILE_PREFIX = "."
 DOWNLOAD_TMP_FILE_SUFFIX = ".nxpart"
+ROOT = Path()
```
[the remark this answers is not in the document]
Comment (LKleinNux) [reply nxdrive/constants.py]: When starting a `Path` object with `/` it assumes we really are on the root of the filesystem. This means `Path("/Users/nuxeo/Nuxeo Drive") / Path("/my workspace")` will evaluate as `Path("/my workspace")`.
This is why I added those functions in `sqlite.py` to add/remove that prefixed `/` to keep the compatibility with old databases but be able to use the paths without it in the rest of the app.

File: nxdrive/client/local_client.py [hunk 6]
```diff
@@ -623,34 +591,33 @@ def onerror(func, path, exc_info):
...
-            if os.path.isfile(os_path):
-                os.unlink(os_path)
-            elif os.path.isdir(os_path):
+            if os_path.is_file():
+                os_path.unlink()
+            elif os_path.is_dir():
                 # Override `onerror` to catch the 1st exception and let other
                 # documents to be deleted.
-                shutil.rmtree(os_path, onerror=onerror)
+                shutil.rmtree(str(os_path), onerror=onerror)
```
[the remark this answers is not in the document]
Comment (LKleinNux) [reply nxdrive/client/local_client.py]: I had trouble with some of `shutil` functions where it did not accept `Path` objects.. I don't know if it was dependent on the version of Python + OS I was on, but I'll check again.

File: nxdrive/engine/processor.py [hunk 7]
```diff
@@ -602,7 +600,9 @@ def _synchronize_locally_created(
                 self._dao.unsynchronize_state(doc_pair, "PARENT_UNSYNC")
                 self._handle_unsynchronized(doc_pair)
                 return
-            raise ParentNotSynced(doc_pair.local_path, doc_pair.local_parent_path)
+            raise ParentNotSynced(
+                str(doc_pair.local_path), str(doc_pair.local_parent_path)
```
[the remark this answers is not in the document]
Comment (LKleinNux) [reply nxdrive/engine/processor.py]: `ParentNotSynced` is used with strings in other places, so I thought I'd keep strings everywhere?

File: nxdrive/engine/watcher/local_watcher.py [hunk 8]
```diff
@@ -1018,9 +1019,9 @@ def handle_watchdog_event(self, evt: FileSystemEvent) -> None:
                             )
                             return
 
-                    rel_parent_path = client.get_path(dirname(src_path))
-                    if rel_parent_path == "":
-                        rel_parent_path = "/"
+                    rel_parent_path = client.get_path(src_path.parent)
+                    if not rel_parent_path:  # Check validity
```
[the remark this answers is not in the document]
Comment (LKleinNux) [reply nxdrive/engine/watcher/local_watcher.py]: Very good question, I guess I forgot to remove it after testing 😬

File: nxdrive/client/local_client.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (BoboTiG) [reply nxdrive/client/local_client.py]: I checked, and on 3.6+ it is OK. If not, for watever reason, you can try to update to Python 3.7.2 first (it should be fine now).

File: nxdrive/constants.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (BoboTiG) [reply nxdrive/constants.py]: OK!

File: nxdrive/client/local_client.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (BoboTiG) [reply nxdrive/client/local_client.py]: The goal is to use `str(Path)` almost never. I will open PRs for third-party modules to accept `Path` objects when possible (like watchdog, xattr, send2trash) and will should do the same for the Python client I think :)

[the remark this answers is not in the document]
Comment (LKleinNux) [reply nxdrive/client/local_client.py]: It happens with `shutil.copy` on macOS with Python 3.6.7 for certain. I'll see with 3.7.2.

File: tests/__init__.py [hunk 9]
```diff
@@ -46,9 +47,19 @@ def requeue_errors(self) -> None:
             doc_pair.error_next_try = 0
 
 
+def _basename(path):
+    """ Patch shutil._basename for pathlib compatibility """
```
Comment (BoboTiG) [inline tests/__init__.py]: Could you add "TODO: remove when https://bugs.python.org/issue32689 is fixed (Python 3.7.3 or newer)"?

**Pull request merged**